lerobot/common/policies/florence/modeling_florence.py

```diff
@@ -0,0 +1,267 @@
+        "horizon" may not the best name to describe what the variable actually means, because this period is
+        actually measured from the first observation which (if `n_obs_steps` > 1) happened in the past.
+        """
+
+        self.eval()
+
+        batch = self.normalize_inputs(batch)
+        if len(self.expected_image_keys) > 0:
+            batch = dict(batch)  # shallow copy so that adding a key doesn't modify the original
+            batch["observation.images"] = torch.stack([batch[k] for k in self.expected_image_keys], dim=-4)
+        # Note: It's important that this happens after stacking the images into a single key.
+        self._queues = populate_queues(self._queues, batch)
+
+        if len(self._queues["action"]) == 0:
+            # stack n latest observations from the queue
+            batch = {k: torch.stack(list(self._queues[k]), dim=1) for k in batch if k in self._queues}
+            actions = self.model(batch)
+
+            # Extract `n_action_steps` steps worth of actions (from the current observation).
+            start = self.config.n_obs_steps - 1
+            end = start + self.config.n_action_steps
+            actions = actions[:, start:end]
+
+            # TODO(rcadene): make above methods return output dictionary?
+            actions = self.unnormalize_outputs({"action": actions})["action"]
+
+            self._queues["action"].extend(actions.transpose(0, 1))
+
+        action = self._queues["action"].popleft()
+        return action
+
+    def forward(self, batch: dict[str, Tensor]) -> dict[str, Tensor]:
+        """Run the batch through the model and compute the loss for training or validation."""
+        batch = self.normalize_inputs(batch)
+        batch = dict(batch)  # shallow copy so that adding a key doesn't modify the original
+        batch["observation.images"] = torch.stack([batch[k] for k in self.expected_image_keys], dim=-4)
+        batch = self.normalize_targets(batch)
+
+        actions_hat = self.model(batch)
+        l1_loss = (
+            F.l1_loss(batch["action"], actions_hat, reduction="none") * ~batch["action_is_pad"].unsqueeze(-1)
+        ).mean()
+
+        return {
+            "loss": l1_loss,
+        }
+
+class FlorenceLinearWrapper(nn.Module):
+    """A wrapper of the Florence2 VLM and a linear action head
+    
+    TODO: the introduction
+    """
+
+    def __init__(self, config: FlorencePolicyConfig):
+        super().__init__()
+
+        self.use_images = any(k.startswith("observation.image") for k in config.input_shapes)
+        self.use_robot_state = "observation.state" in config.input_shapes
+        self.use_env_state = "observation.environment_state" in config.input_shapes
+        if self.use_images == False:
+            raise Exception("modeling_florence_policy.py: you should use florence policy with image input")
+        if config.crop_shape is not None:
```

Comment:
Does Florence do random cropping or is this just us adding it?

Reply:
Hi @alexander-soare, good question!. Florence2 does not have random cropping, please check [here](https://huggingface.co/microsoft/Florence-2-base/blob/main/preprocessor_config.json). It actually uses the [CLIPImageProcessor](https://github.com/huggingface/transformers/blob/2a5a6ad18aa22e98429bb5ecb880660328030ea0/src/transformers/models/clip/image_processing_clip.py#L52), and the process is:

1. Resize the image to (768, 768)
2. Rescale the image to 0-1 (It seems that you alreandy do it in LeRobot, but I cannot find the location. Could you please send me a link?)
3. Normalize it according to `image_mean: [0.485, 0.456, 0.406], image_std: [0.229, 0.224, 0.225]`

You are right! I shall keep it as same as the original setting, except for the image resolution.

Reply:
Regarding 2, it's here: https://github.com/huggingface/lerobot/blob/03d647269eff6d76df14c84159adc7624a582762/lerobot/common/datasets/video_utils.py#L159

Reply:
I see, thanks!